app/multitenant/aws_collector.go

```diff
@@ -348,25 +354,89 @@ func (c *awsCollector) getReports(ctx context.Context, reportKeys []string) ([]r
 	return reports, nil
 }
 
+/*
+S3 stores original reports from one probe at the timestamp they arrived at collector.
+Collector also sends every report to memcached.
+The in-memory cache stores:
+ - individual reports deserialised, under S3 key for report
+ - sets of reports in interval [t,t+3) merged, under key "instance:t"
+   - so to check the cache for reports from 14:31:00 to 14:31:15 you would request 5 keys 3 seconds apart
+*/
+
 func (c *awsCollector) Report(ctx context.Context, timestamp time.Time) (report.Report, error) {
 	span, ctx := opentracing.StartSpanFromContext(ctx, "awsCollector.Report")
 	defer span.Finish()
-	reportKeys, err := c.getReportKeys(ctx, timestamp)
+	userid, err := c.cfg.UserIDer(ctx)
 	if err != nil {
 		return report.MakeReport(), err
 	}
-	span.LogFields(otlog.Int("keys", len(reportKeys)), otlog.String("timestamp", timestamp.String()))
-	log.Debugf("Fetching %d reports to %v", len(reportKeys), timestamp)
-	reports, err := c.getReports(ctx, reportKeys)
+	end := timestamp
+	start := end.Add(-c.cfg.Window)
+	reportKeys, err := c.getReportKeys(ctx, userid, start, end)
 	if err != nil {
 		return report.MakeReport(), err
 	}
+	span.LogFields(otlog.Int("keys", len(reportKeys)), otlog.String("timestamp", timestamp.String()))
 
+	var reports []report.Report
+	// Fetch a merged report for each time quantum covering the window
+	startTS, endTS := start.UnixNano(), end.UnixNano()
+	ts := startTS - (startTS % reportQuantisationInterval.Nanoseconds())
+	for ; ts+(reportQuantisationInterval+gracePeriod).Nanoseconds() < endTS; ts += reportQuantisationInterval.Nanoseconds() {
+		quantumReport, err := c.reportForQuantum(ctx, userid, reportKeys, ts)
+		if err != nil {
+			return report.MakeReport(), err
+		}
+		reports = append(reports, quantumReport)
+	}
```

Comment:
I was wondering whether that entire loop could operate on `time.Time`, converting to unix nanos only in the call to (or even inside) `reportForQuantum`.

Reply:
It’s the modulus that scares me. 
I’ve made one bug that way already this year.